File: scripts/pareto_diag_sample.r
library(posterior)

N=1000
phi=0.3
set.seed(6534)
dr<-array(data=replicate(4,as.numeric(arima.sim(n = N,
                                                list(ar = c(phi)),
                                                sd = sqrt((1-phi^2))))),
         dim=c(N,4,1)) |>
  as_draws_df() |>
  set_variables('xn')


drt <- dr |>
  mutate_variables(xt3=qt(pnorm(xn), df=3),
                   xt2_5=qt(pnorm(xn), df=2.5),
                   xt2=qt(pnorm(xn), df=2),
                   xt1_5=qt(pnorm(xn), df=1.5),
                   xt1=qt(pnorm(xn), df=1))

saveRDS(drt, "../models/prerun/drt.rds")
write.csv(as_draws_df(drt), "../models/prerun/drt.csv", row.names = FALSE)

File: scripts/pareto_diag_sample.py
import pandas as pd
import numpy as np
import arviz as az
import xarray as xr

df = pd.read_csv("../models/prerun/drt.csv")

variables = ["xn", "xt3", "xt2_5", "xt2", "xt1_5", "xt1"]
n_chains = df[".chain"].nunique()
n_draws  = df[".iteration"].nunique()

# Reshape each variable to (chain, draw)
data = {}
for v in variables:
    data[v] = (
        df.sort_values([".chain", ".iteration"])
        .groupby(".chain")[v]
        .apply(np.array)
        .values
        .tolist()  # list of arrays, one per chain
    )
    data[v] = np.array(data[v])  # shape: (chain, draw)

# Build xarray Dataset then DataTree
ds = xr.Dataset(
    {v: (["chain", "draw"], data[v]) for v in variables},
    coords={"chain": range(n_chains), "draw": range(n_draws)}
)

drt = az.from_dict({"posterior": ds})
drt.to_netcdf("../models/prerun/drt.nc")